File: web/src/lib/bm25.test.ts
import { describe, it, expect } from 'vitest';
import { bm25Score, bm25RankTokens, hybridRankTokens } from './bm25.js';
// Each hybrid test gets its own empty cache to avoid cross-test cache pollution.
const freshCache = () => new Map<string, Float32Array>();
import type { TokenMeta } from './grammar.js';

function makeMeta(token: string, label: string, description: string, heuristics: string[] = [], distinctions: string[] = []): TokenMeta {
	return {
		token,
		label,
		description,
		guidance: '',
		use_when: '',
		kanji: '',
		category: '',
		routing_concept: '',
		metadata: { definition: description, heuristics, distinctions: distinctions.map(d => ({ token: d, note: '' })) }
	} as unknown as TokenMeta;
}

const corpus: TokenMeta[] = [
	makeMeta('gate', 'Evaluation before implementation', 'Enforces that a FAIL observation per assertion exists before implementation proceeds.', ['TDD', 'red-green-refactor'], ['chain', 'atomic']),
	makeMeta('chain', 'Derivation chain enforcement', 'Each step must reproduce its predecessor actual output verbatim before proceeding.', ['reasoning chain', 'continuity'], ['gate']),
	makeMeta('ground', 'Establish correctness through evidence', 'The gap between apparent completion and actual completion is the optimizer attack surface.', ['enforcement process', 'governing artifact'], ['verify']),
	makeMeta('probe', 'Structural diagnosis', 'Surface hidden assumptions and root causes before planning.', ['root cause', 'diagnose']),
];

// Assertion A: bm25Score returns score > 0 for tokens matching any query term (OR semantics)
describe('bm25Score', () => {
	it('A1: returns positive score for a token matching a query term', () => {
		const scores = bm25Score(corpus, 'TDD');
		expect(scores.get('gate')).toBeGreaterThan(0);
	});

	it('A2: returns 0 (or no entry) for a token not matching any query term', () => {
		const scores = bm25Score(corpus, 'TDD');
		expect(scores.get('probe') ?? 0).toBe(0);
	});

	it('A3: OR semantics — token matching one of two words scores positively', () => {
		const scores = bm25Score(corpus, 'verbatim xyzzzz');
		// 'chain' has 'verbatim' in its description; 'xyzzzz' matches nothing
		expect(scores.get('chain')).toBeGreaterThan(0);
	});
});

// Assertion B: bm25RankTokens returns tokens sorted by score descending
describe('bm25RankTokens — ordering', () => {
	it('B1: top result is most relevant token for specific query', () => {
		const ranked = bm25RankTokens(corpus, 'TDD assertion');
		expect(ranked.length).toBeGreaterThan(0);
		expect(ranked[0].token.token).toBe('gate');
	});

	it('B2: results are sorted by score descending', () => {
		const ranked = bm25RankTokens(corpus, 'predecessor output');
		// chain has 'predecessor' and 'output' in description — should rank high
		for (let i = 0; i < ranked.length - 1; i++) {
			expect(ranked[i].score).toBeGreaterThanOrEqual(ranked[i + 1].score);
		}
	});
});

// Assertion S: Porter stemming — inflected query form matches doc containing root form
describe('bm25Score — stemming', () => {
	it('S1: query with inflected form matches doc containing root form', () => {
		const stemCorpus = [
			makeMeta('routeplanner', 'Route Planner', 'optimizes route planning for navigation'),
			makeMeta('other', 'Other', 'completely unrelated content about databases'),
		];
		const scores = bm25Score(stemCorpus, 'routing');
		expect(scores.get('routeplanner')).toBeGreaterThan(0);
	});
});

// Assertion C: multi-word query where no single token matches all words still returns results
describe('bm25RankTokens — multi-word OR fallback', () => {
	it('C1: query with terms spread across different tokens returns results for each', () => {
		// 'TDD' matches gate; 'optimizer' matches ground; no single token has both in heuristics
		const ranked = bm25RankTokens(corpus, 'TDD optimizer');
		const tokenNames = ranked.map((r) => r.token.token);
		expect(tokenNames).toContain('gate');
		expect(tokenNames).toContain('ground');
	});
});

// Assertion H: hybridRankTokens — embedding-weighted hybrid search
describe('hybridRankTokens', () => {
	it('H1: with null embedder degrades to BM25-only ordering', async () => {
		const hybrid = await hybridRankTokens(corpus, 'TDD assertion', null, freshCache());
		const bm25Only = bm25RankTokens(corpus, 'TDD assertion');
		expect(hybrid.map(r => r.token.token)).toEqual(bm25Only.map(r => r.token.token));
	});

	it('H2: embedder is called with token doc text and cosine similarity influences ranking', async () => {
		// Stub returns [1,0] for the query and for text containing "root cause" (probe's heuristic),
		// [0,1] otherwise. probe should rank first despite zero BM25 overlap with the query.
		const stubEmbedder = async (text: string) =>
			(text === 'unrelated query xyz' || text.includes('root cause'))
				? new Float32Array([1, 0])
				: new Float32Array([0, 1]);
		const hybrid = await hybridRankTokens(corpus, 'unrelated query xyz', stubEmbedder, freshCache());
		expect(hybrid.length).toBeGreaterThan(0);
		expect(hybrid[0].token.token).toBe('probe');
	});

	it('H4: embedder is NOT called for tokens that have pre-computed metadata.embedding; only the query is embedded', async () => {
		// All tokens have pre-computed embeddings: probe=[1,0], others=[0,1].
		const corpusWithEmb = corpus.map(t => ({
			...t,
			metadata: { ...t.metadata, embedding: t.token === 'probe' ? [1, 0] : [0, 1] }
		})) as unknown as TokenMeta[];

		let embedderCallCount = 0;
		const stubEmbedder = async (text: string) => {
			embedderCallCount++;
			// Query returns [1,0]; token texts would return [0,1] if mistakenly called
			return text === 'unrelated query xyz' ? new Float32Array([1, 0]) : new Float32Array([0, 1]);
		};
		const hybrid = await hybridRankTokens(corpusWithEmb, 'unrelated query xyz', stubEmbedder, freshCache());
		// Only the query should have been embedded (1 call), not any token texts
		expect(embedderCallCount).toBe(1);
		// probe's pre-computed [1,0] matches query [1,0] → cosine=1, ranks first
		expect(hybrid[0].token.token).toBe('probe');
	});

	it('H3: embedder is called to embed token doc text (description+heuristics+distinctions+routing_concept) when no pre-computed embedding exists', async () => {
		// corpus tokens have NO .embedding field in metadata.
		// We use a stub embedder that returns a high-similarity vector only for
		// text containing "root cause" (the probe heuristic). Query is unrelated to
		// BM25 terms so BM25 alone would not rank probe first.
		const embeddedTexts: string[] = [];
		const stubEmbedder = async (text: string) => {
			embeddedTexts.push(text);
			// Return high similarity for text mentioning "root cause"
			return text.includes('root cause') ? new Float32Array([1, 0]) : new Float32Array([0, 1]);
		};
		// Query vector matches [1, 0] → probe should win over gate/chain/ground on cosine
		const queryEmbedder = async (text: string) => {
			if (text === 'unrelated query xyz') return new Float32Array([1, 0]);
			return stubEmbedder(text);
		};
		const hybrid = await hybridRankTokens(corpus, 'unrelated query xyz', queryEmbedder, freshCache());
		// Embedder must have been called for token descriptions (not just the query)
		expect(embeddedTexts.length).toBeGreaterThan(1);
		// probe should rank first because its doc text contains "root cause"
		expect(hybrid[0].token.token).toBe('probe');
	});
});

File: internal/barcli/bm25_test.go
package barcli

import (
	"strings"
	"testing"
)

// TestBM25ScoresStemming verifies that a query with an inflected form matches
// a document containing only the root form, via Porter stemming.
func TestBM25ScoresStemming(t *testing.T) {
	docs := []tokenDoc{
		{id: "a", title: "route planner", body: "optimizes route planning for navigation"},
		{id: "b", title: "database", body: "completely unrelated content about databases"},
	}
	scores := bm25Scores(docs, "routing")
	if scores["a"] == 0 {
		t.Errorf("bm25Scores: query 'routing' should match doc containing 'route' via stemming, got score 0")
	}
}


// TestBuildTokenDocsIncludesAxisHeuristics verifies buildTokenDocs includes axis-level
// heuristics in each token's BM25 body so queries like "reasoning approach" match method tokens.
func TestBuildTokenDocsIncludesAxisHeuristics(t *testing.T) {
	g := &Grammar{
		Axes: AxisSection{
			Metadata: map[string]map[string]TaskMetadata{
				"method": {"flow": {Definition: "Linear stage sequencing"}},
			},
			AxisHeuristics: map[string][]string{
				"method": {"reasoning approach", "how to think"},
			},
		},
	}
	docs := buildTokenDocs(g, "")
	var found *tokenDoc
	for i := range docs {
		if docs[i].id == "method:flow" {
			found = &docs[i]
			break
		}
	}
	if found == nil {
		t.Fatal("doc method:flow not found")
	}
	if !strings.Contains(found.body, "reasoning approach") {
		t.Errorf("buildTokenDocs: body for method:flow should contain axis heuristic 'reasoning approach', got: %q", found.body)
	}
}

// TestBuildTokenDocsIncludesAxisDescription verifies buildTokenDocs includes the axis-level
// description in each token's BM25 body.
func TestBuildTokenDocsIncludesAxisDescription(t *testing.T) {
	g := &Grammar{
		Axes: AxisSection{
			Metadata: map[string]map[string]TaskMetadata{
				"method": {"flow": {Definition: "Linear stage sequencing"}},
			},
			AxisDescriptions: map[string]string{
				"method": "Reasoning approach — how to think through the problem",
			},
		},
	}
	docs := buildTokenDocs(g, "")
	var found *tokenDoc
	for i := range docs {
		if docs[i].id == "method:flow" {
			found = &docs[i]
			break
		}
	}
	if found == nil {
		t.Fatal("doc method:flow not found")
	}
	if !strings.Contains(found.body, "Reasoning approach") {
		t.Errorf("buildTokenDocs: body for method:flow should contain axis description, got: %q", found.body)
	}
}

// TestBuildTokenDocsIncludesRoutingConcept verifies buildTokenDocs includes routing_concept
// in the BM25 body for axis tokens.
func TestBuildTokenDocsIncludesRoutingConcept(t *testing.T) {
	g := &Grammar{
		Axes: AxisSection{
			Metadata: map[string]map[string]TaskMetadata{
				"method": {"flow": {Definition: "Linear stage sequencing"}},
			},
			RoutingConcept: map[string]map[string]string{
				"method": {"flow": "sequence steps in a pipeline"},
			},
		},
	}
	docs := buildTokenDocs(g, "")
	var found *tokenDoc
	for i := range docs {
		if docs[i].id == "method:flow" {
			found = &docs[i]
			break
		}
	}
	if found == nil {
		t.Fatal("doc method:flow not found")
	}
	if !strings.Contains(found.body, "sequence steps in a pipeline") {
		t.Errorf("buildTokenDocs: body for method:flow should contain routing_concept, got: %q", found.body)
	}
}

// TestBM25ScoresRRFFormula verifies that bm25ScoresRRF computes Σ 1/(60+rank_f)
// across per-field corpora: a token ranking 1st in title only scores 1/61.
func TestBM25ScoresRRFFormula(t *testing.T) {
	// doc "a" has query word in title only; doc "b" has it in definition only.
	docs := []tokenDoc{
		{id: "a", title: "widget", heuristics: "unrelated", distinctions: "unrelated", definition: "unrelated"},
		{id: "b", title: "unrelated", heuristics: "unrelated", distinctions: "unrelated", definition: "widget"},
	}
	scores := bm25ScoresRRF(docs, "widget")
	if scores["a"] == 0 {
		t.Error("bm25ScoresRRF: doc with query in title should have non-zero RRF score")
	}
	if scores["b"] == 0 {
		t.Error("bm25ScoresRRF: doc with query in definition should have non-zero RRF score")
	}
	// Both rank 1st in their respective field and are absent from the other three.
	// Each should score exactly 1/(60+1) = 1/61.
	want := 1.0 / 61.0
	const eps = 1e-9
	if got := scores["a"]; got < want-eps || got > want+eps {
		t.Errorf("bm25ScoresRRF: doc 'a' (title-only match) score = %v, want %v", got, want)
	}
	if got := scores["b"]; got < want-eps || got > want+eps {
		t.Errorf("bm25ScoresRRF: doc 'b' (definition-only match) score = %v, want %v", got, want)
	}
}

// TestBM25ScoresRRFExclusion verifies tokens absent from all field rankings are excluded.
func TestBM25ScoresRRFExclusion(t *testing.T) {
	docs := []tokenDoc{
		{id: "match", title: "widget", heuristics: "widget", distinctions: "widget", definition: "widget"},
		{id: "nomatch", title: "unrelated", heuristics: "unrelated", distinctions: "unrelated", definition: "unrelated"},
	}
	scores := bm25ScoresRRF(docs, "widget")
	if _, ok := scores["nomatch"]; ok {
		t.Error("bm25ScoresRRF: token absent from all field rankings should be excluded from results")
	}
	// Also verify no zero-score entry leaks into the result map.
	for id, s := range scores {
		if s == 0 {
			t.Errorf("bm25ScoresRRF: result map contains zero-score entry for %q", id)
		}
	}
}

// TestBM25ScoresRRFFieldIsolation verifies each field corpus contains only its own text.
// Doc "a" has "alpha" only in heuristics; doc "b" has empty heuristics but "alpha" must not
// bleed from doc "a"'s heuristics into the definition corpus where doc "b" ranks.
func TestBM25ScoresRRFFieldIsolation(t *testing.T) {
	docs := []tokenDoc{
		// "alpha" in heuristics only; definition is unique "zeta"
		{id: "a", title: "unrelated", heuristics: "alpha", distinctions: "unrelated", definition: "zeta"},
		// "alpha" absent from all fields; definition is unique "zeta" too — so if heuristics bled into definition, both would rank in definition corpus
		{id: "b", title: "unrelated", heuristics: "", distinctions: "unrelated", definition: "zeta"},
	}
	scores := bm25ScoresRRF(docs, "alpha")
	if scores["a"] == 0 {
		t.Error("bm25ScoresRRF field isolation: doc with query in heuristics should score non-zero")
	}
	if _, ok := scores["b"]; ok {
		t.Error("bm25ScoresRRF field isolation: doc without query in any field should be excluded even if other fields share text")
	}
}

// TestBM25TokenizeStem verifies that bm25Tokenize applies Porter stemming.
func TestBM25TokenizeStem(t *testing.T) {
	cases := []struct {
		in   string
		want string
	}{
		{"routing", "rout"},
		{"routes", "rout"},
	}
	for _, c := range cases {
		got := bm25Tokenize(c.in)
		if len(got) != 1 || got[0] != c.want {
			t.Errorf("bm25Tokenize(%q) = %v, want [%s]", c.in, got, c.want)
		}
	}
}
